cpp/farm_ng/core/prototools/event_log_reader.cpp

```diff
@@ -78,34 +91,103 @@ class EventLogReaderBinaryImpl : public EventLogReaderImpl {
   std::ifstream in;
 };
 
-namespace {
-void resetImpl(
-    std::unique_ptr<EventLogReaderImpl>& impl,
-    std::filesystem::path const& log_path) {
-  impl = std::make_unique<EventLogReaderBinaryImpl>(log_path);
+EventLogPos::EventLogPos(
+    core::proto::Event event,
+    std::streampos pos,
+    std::weak_ptr<EventLogReaderImpl> log)
+    : event_(std::move(event)), pos_(pos), log_(log) {}
+
+core::proto::Event const& EventLogPos::event() const { return event_; }
+
+std::string EventLogPos::readPayload() const {
+  std::shared_ptr<EventLogReaderImpl> log = log_.lock();
+  FARM_CHECK(!!log, "Log closed: {}", event_.ShortDebugString());
+  return log->readPayload(event_, pos_);
 }
-}  // namespace
 
-EventLogReader::EventLogReader(std::filesystem::path const& log_path) {
-  resetImpl(impl_, log_path);
+std::vector<EventLogPos> const& EventLogReaderImpl::getIndex() {
+  if (index_.empty()) {
+    reset();
+    while (true) {
+      try {
```

Comment:
nit: suggest using expected instead.

Reply:
ok, for now I'll go with exceptions based interface for this library.  There are a few places in the stack that can error out, its a bit easy to code the happy path and let the exceptions propagate.